builder.rb

```diff
@@ -33,12 +34,53 @@ def build
       puts "\e[34m#{build_command}\e[0m"
       puts
 
-      raise 'Build failed' unless system(build_command.join(' '))
+      if ([MDTOOL_PATH, 'build', 'archive'] & build_command).any?
+        run_mdtool_in_diagnostic_mode(build_command)
+      else
+        raise 'Build failed' unless system(build_command.join(' '))
+      end
     end
 
     @generated_files = @analyzer.collect_generated_files(@configuration, @platform, @project_type_filter)
   end
 
+  # README:
+  # This method will run `mdtool build` in diagnostic mode.
+  # If mdtool will hang trying to load projects its process will be killed
+  # and stack trace for each thread of mdtool will be printed to stdout.
+  # Issue on Bugzilla: https://bugzilla.xamarin.com/show_bug.cgi?id=42378
+
+  # List of things to be removed as as soon as #42378 will be resolved:
+  # 1) run_mdtool_in_diagnostic_mode method
+  # 2) hijack_process method
+  # 3) MDTOOL_PATH constant in Builder class
+  # 4) Entire Timer class
+  # 5) if/else logic in Builder.build
+
+  MDTOOL_PATH = "\"/Applications/Xamarin Studio.app/Contents/MacOS/mdtool\""
+
+  def run_mdtool_in_diagnostic_mode(mdtool_build_command)
+    pipe = nil
+
+    timer = Timer.new(900) { # 15 minutes timeout
```

Comment:
i think we should decrease the timeout, our free users have 10 min (600 sec) for a whole build.
i recommand to set timeout to ~ 5 min (300 sec), otherwise, free user's builds will be aborted before the stack trace printed out.
also as i seen till now, loading the project does not takes to much time.